core/modules/parser/QSMySqlListener.cc

```diff
@@ -2725,47 +2707,50 @@ class LogicalExpressionAdapter :
-        for (auto term : _terms) {
-            if (false == _logicalOperator->merge(*term)) {
-                if (isOr) {
-                    _logicalOperator->addBoolTerm(make_shared<query::AndTerm>(term));
-                } else {
-                    _logicalOperator->addBoolTerm(term);
+        ASSERT_EXECUTION_CONDITION(_logicalOperatorIsSet, "logicalOperator is not set.", _ctx);
+        shared_ptr<query::LogicalTerm> logicalTerm;
+        switch (_logicalOperatorType) {
+            case LogicalOperatorCBH::AND: {
+                logicalTerm = make_shared<query::AndTerm>();
+                for (auto term : _terms) {
+                    if (false == logicalTerm->merge(*term)) {
+                        logicalTerm->addBoolTerm(term);
+                    }
+                }
+                break;
+            }
+            case LogicalOperatorCBH::OR:
+            {
+                auto orTerm = make_shared<query::OrTerm>();
+                logicalTerm = orTerm;
+                for (auto term : _terms) {
+                    if (false == logicalTerm->merge(*term)) {
+                        logicalTerm->addBoolTerm(make_shared<query::AndTerm>(term));
+                    }
                 }
+                break;
             }
+
+            default:
+                ASSERT_EXECUTION_CONDITION(false, "unhandled logical operator.", _ctx);
         }
-        lockedParent()->handleLogicalExpression(_logicalOperator, _ctx);
+        lockedParent()->handleLogicalExpression(logicalTerm, _ctx);
     }
 
     string name() const override { return getTypeName(this); }
 
 private:
-    void _setLogicalOperator(shared_ptr<query::LogicalTerm> const & logicalTerm) {
-        ASSERT_EXECUTION_CONDITION(nullptr == _logicalOperator,
-                "logical operator must be set only once.", _ctx);
-        _logicalOperator = logicalTerm;
-    }
-
     friend ostream& operator<<(ostream& os, const LogicalExpressionAdapter& logicaAndlExpressionAdapter);
 
-    // a qserv restrictor fucntion can be the left side of a predicate (currently it can only be the left
-    // side; that is to say, it can only be the first term in the WHERE clause. If `handleQservFunctionSpec`
-    // is called and _leftTerm is null (as well as _rightTerm and _logicalOperator, then _leftHandled is set
-    // to true to indicate that the left term has been handled. This allows onExit to put only one term into
-    // the logicalOperator and know that it was ok (the qserv IR accepts an AndTerm with only one factor).
-    // This mechanism does not fully proect against qserv restrictors that may be the left side of a
-    // subsequent logical expression. TBD if that's really an issue.
     vector<shared_ptr<query::BoolTerm>> _terms;
-    shared_ptr<query::LogicalTerm> _logicalOperator;
+    LogicalOperatorCBH::OperatorType _logicalOperatorType;
+    bool _logicalOperatorIsSet {false};
```

Comment:
this is in the class declaration, not a function decl, so that wouldn't work here.

Reply:
@andy-slac is this still true? I heard rumor that this was because of a version of the gcc compiler that we no longer use (4.8?)

Reply:
I think it still applies: https://developer.lsst.io/cpp/style.html#a-constructor-calls-should-use-c-98-syntax-when-possible
Old compiler had some issues, but even with new compiler it's better to avoid possible confusion.